doc: add pages about data and key storage (#22683)
nrfconnect/sdk-nrf: doc/nrf/security/key_storage.rst

This library is leveraging either Zephyr's :ref:`Settings subsystem <zephyr:settings_api>` or :ref:`PSA's Internal Trusted Storage (ITS) <ug_psa_certified_api_overview_secstorage>` to store credentials.

This option offers the following features:

* PSA Protected Storage backend - Benefits from encryption and integrity protection.
* Easy-to-use API - Simplified interface for Wi-Fi credential management.
* Multiple networks - Support for storing multiple Wi-Fi network credentials.

Configuration
-------------

To use the Wi-Fi credentials library, enable the :kconfig:option:`CONFIG_WIFI_CREDENTIALS` Kconfig option.
For more configuration options, see the :ref:`lib_wifi_credentials` page.

Usage example
-------------

See the :ref:`Wi-Fi Bluetooth LE based provision sample <wifi_provisioning>` for an example.

Additional security considerations
**********************************

.. _key_storage_flash_protection:

Hardware Flash Write Protection
===============================

The |NCS| provides the :ref:`fprotect driver <fprotect_readme>` for hardware-level flash protection.

> This driver uses underlying hardware peripherals (BPROT, ACL, or SPU) to protect flash regions from unauthorized writes or reads.

Author: Are you sure? I can't find anything about RRAMC in the sources.

> This driver uses underlying hardware peripherals (BPROT, ACL, or SPU) to protect flash regions from unauthorized writes or reads.

Author: Expanded the page with information about RRAMC and ZMS.

Use fprotect to protect stored keys from modification, secure bootloader regions, and implement additional layers of security.

Configuration
-------------

To use the fprotect driver, enable the :kconfig:option:`CONFIG_FPROTECT` Kconfig option.
For more information, see the :ref:`fprotect_readme` page.

.. _key_storage_approtect:

> Access Port Protection (AP-Protect)

Author: It's already short :)

.. _key_storage:

Key storage options
###################

.. contents::
   :local:
   :depth: 2

Cryptographic keys are critical security assets that require secure storage and management.

This page provides an overview of the available key storage alternatives in the |NCS|.

For information about storing general data, see :ref:`data_storage`.

General key storage recommendations
***********************************

When implementing key storage in your application:

* Use the PSA Crypto API for all cryptographic operations.
* Generate keys on-device when possible rather than importing them.
* Leverage hardware features like Hardware Unique Key (HUK) and Key Management Unit (KMU) when available.
* Enable APPROTECT in production to prevent debug access.
* Use persistent keys through the PSA Crypto API when you need to store key material.
* Implement proper key provisioning workflows for production devices.

.. _key_storage_alternatives:

> Key storage alternatives

Author: That requires changing the page title, which I now did. I don't like neither `options` nor `alternatives`, truth be told.

> Key storage alternatives

Reviewer: Ah sorry I looked at the diff the wrong way arond. Indeed. I do like the new title though.

-------------

For information about how to configure the PSA Crypto API in the |NCS|, see :ref:`psa_crypto_support`.

Usage example
-------------

For usage examples, see :ref:`crypto_samples`, and specifically the :ref:`Persistent key usage <crypto_persistent_key>` sample.

.. _key_storage_huk:

Hardware Unique Key (HUK)
=========================

Hardware Unique Keys (HUKs) are device-specific keys that you can use for key derivation.
HUKs provide a foundation for creating device-unique cryptographic keys without storing key material in software.

This option offers the following features:

* Device uniqueness - Each device has a unique set of HUK values.
* Hardware protected - Keys are stored in hardware-protected memory.
* Key derivation - HUKs are used to derive other keys rather than directly perform cryptographic operations.
  The derived keys are then passed to the application.

See the :ref:`lib_hw_unique_key` library for the list of supported HUKs in the |NCS|.

Device support
--------------

Different nRF devices support different types of HUKs.
See :ref:`lib_hw_unique_key` for device-specific support information.

Usage
-----

You cannot use HUKs directly.

> Use the :ref:`PSA Protected Storage APIs <ug_psa_certified_api_overview_secstorage>` to use them.

Author: How about:
```
You cannot use HUKs directly for non-volatile key derivation.
[... <skip to the end of the section>]
.. note::
   You can still use the HUK library directly to derive volatile keys for testing and optimization.
```

.. _key_storage_kmu:

Key Management Unit (KMU)
=========================

The Key Management Unit (KMU) is a hardware peripheral for secure key storage available on select nRF devices.
It provides hardware-level protection for cryptographic keys.

This option offers the following features:

* Hardware security - Keys are stored in dedicated hardware.
* CPU isolation - CryptoCell and CRACEN can use KMU keys without CPU access to key material.
* Metadata only - CPU only knows key metadata, not key material.
* Multiple key slots - Supports storing multiple keys separately.

Device support
--------------

KMU is available on the following devices:

* `nRF5340 <nRF5340 Key management unit_>`_
* nRF91 Series devices, such as `nRF9160 <nRF9160 Key management unit_>`_ or `nRF9151 <nRF9151 Key management unit_>`_
* nRF54L Series devices, such as `nRF54L15 <nRF54L15 Key management unit_>`_

  .. note::
     nRF54L devices are equipped with a KMU that works with the CRACEN peripheral.
     For more information, see :ref:`ug_nrf54l_developing_basics_kmu`.

Key storage
-----------

The KMU stores various types of keys:

* Hardware Unique Keys (MKEK, MEXT)

> * Initial Attestation Key (for devices other than nRF54L Series, where the IAK is derived)

Reviewer: The same could be said of the HUKs I believe.

Author: You sound unsure, so I did not make any edit.

Reviewer: On this line where I commented you are documenting this very thing (that the HUKs are stored in the KMU, and in addition I pointed out that it's not the case for nRF54 Series): https://github.com/nrfconnect/sdk-nrf/pull/22683#discussion_r2200866087

* Easy-to-use API - Simplified interface for Wi-Fi credential management.
* Multiple networks - Support for storing multiple Wi-Fi network credentials.

Configuration
-------------

To use the Wi-Fi credentials library, enable the :kconfig:option:`CONFIG_WIFI_CREDENTIALS` Kconfig option.
For more configuration options, see the :ref:`lib_wifi_credentials` page.

Usage example
-------------

See the :ref:`Wi-Fi Bluetooth LE based provision sample <wifi_provisioning>` for an example.

Additional security considerations
**********************************

.. _key_storage_iak:

Initial Attestation Key (IAK) / Identity Key
============================================

The :ref:`Initial Attestation Key (IAK) <lib_identity_key>`, also known as Identity Key, is required by PSA Certified Security Framework for :ref:`ug_psa_certified_api_overview_attestation`.
This key is used to sign attestation tokens that prove the device's identity and integrity state.

This option offers the following features:

* Device identity - Provides cryptographic proof of device identity.
* Attestation support - Required for PSA Attestation implementation.

> * Hardware protection - Encrypted using MKEK and either stored in :ref:`key_storage_kmu` (for devices other than nRF54L Series) or derived every time (for nRF54L Series devices).

Reviewer: I think that we could just say generally for both cases that the IAK is only accessible by HW. And I believe it's `Encrypted using MKEK` only if it's stored, not if it's derived every time.

Usage example
-------------

For usage examples, see :ref:`crypto_samples`, and specifically the :ref:`Persistent key usage <crypto_persistent_key>` sample.

.. _key_storage_huk:

Hardware Unique Key (HUK)
=========================

Hardware Unique Keys (HUKs) are device-specific keys that you can use for key derivation.
HUKs provide a foundation for creating device-unique cryptographic keys without storing key material in software.

This option offers the following features:

* Device uniqueness - Each device has a unique set of HUK values.
* Hardware protection - Keys are stored in hardware-protected memory.
* Key derivation - HUKs are used to derive other keys rather than directly perform cryptographic operations.
  The derived keys are then passed to the application.

See the :ref:`lib_hw_unique_key` library for the list of supported HUKs in the |NCS|.

Device support
--------------

Different Nordic Semiconductor devices support different types of HUKs.
See :ref:`lib_hw_unique_key` for device-specific support information.

Usage
-----

You cannot use HUKs directly.

You can however use the :ref:`lib_hw_unique_key` library to derive keys from HUKs using known labels.

> If you use HUKs indirectly through PSA Secure Storage without TF-M, for example through the :ref:`trusted_storage_readme` library, the HUK library will be used to derive keys, which will then be used to encrypt storage entries.

Reviewer: Agreeing except the first comma you are adding.

To use the Wi-Fi credentials library, enable the :kconfig:option:`CONFIG_WIFI_CREDENTIALS` Kconfig option.
For more configuration options, see the :ref:`lib_wifi_credentials` page.

Usage example
-------------

See the :ref:`Wi-Fi Bluetooth LE based provision sample <wifi_provisioning>` for an example.

Additional considerations
*************************

The |NCS| includes several additional features that are related to cryptographic key storage.

.. _key_storage_iak:

Initial Attestation Key (IAK) / Identity Key
============================================

The :ref:`Initial Attestation Key (IAK) <lib_identity_key>`, also known as Identity Key, is required by PSA Certified Security Framework for :ref:`ug_psa_certified_api_overview_attestation`.
This key is used to sign attestation tokens that prove the device's identity and integrity state.

This option offers the following features:

* Device identity - Provides cryptographic proof of device identity.
* Attestation support - Required for PSA Attestation implementation.
* Hardware protection - Only accessible by hardware.
  Encrypted using MKEK when stored in :ref:`key_storage_kmu` for devices other than nRF54L Series.

>   Derived every time for nRF54L Series devices.

Reviewer: Not data. Keys.

>   For devices other than the nRF54L Series, data is encrypted using MKEK and stored in :ref:`key_storage_kmu`.
>   For nRF54L Series devices, the keys are derived each time.

Reviewer: Actually we're talking about a single key (the IAK).
Also:
```suggestion
  For devices other than the nRF54L Series, the key is encrypted using MKEK and stored in :ref:`key_storage_kmu`.
  For nRF54L Series devices, the key is derived each time.
```

* Enable APPROTECT in production to prevent debug access.
* Use persistent keys.
* Implement proper key provisioning workflows for production devices.

.. _key_storage_alternatives:

Key storage alternatives
************************

Each key storage alternative offers different security features and use cases.

.. _key_storage_psa_crypto:

PSA Crypto API
==============

The :ref:`PSA Crypto API <ug_psa_certified_api_overview_crypto>` is the recommended approach for cryptographic key management in the |NCS|.
It provides a standardized, secure interface for key operations that abstracts the underlying hardware and software implementations.

This option offers the following features:

* Secure key handling - Keys are managed securely without exposing key material.
* Hardware acceleration - Utilizes available cryptographic hardware when possible.
* Standardized interface - Follows PSA Certified standards for interoperability.
* Key isolation - Keys can be isolated in the Secure Processing Environment (SPE) when using :ref:`Trusted Firmware-M <ug_tfm>`.

Key generation and import
-------------------------

> The PSA Crypto API supports two methods for managing keys:

Reviewer: ```suggestion
The PSA Crypto API supports two methods for using keys:
```

Key generation
  Use :c:func:`psa_generate_key` to create random keys.
  Generated keys can have one of the following lifetimes:

  * Volatile - Stored in RAM, lost on reset.

>   * Persistent - Stored in PSA Internal Trusted Storage, retained during reboots.

Reviewer: Well actually volatile or persistent is not a property only for generated keys. I'm pretty sure it applies just the same to imported keys.

Key import
  Use :c:func:`psa_import_key` to import existing key material.
  After import, the key is referenced by ID rather than direct access to key material.

.. note::
   Generated keys are more secure than imported keys since they are never exposed to the Non-Secure Processing Environment (NSPE).
   When the key is imported from the NSPE, it temporarily resides in an untrusted environment.

>    Whenever possible, it is safer to generate new keys without exposing them to the NSPE.

Reviewer: This note is only valid when using TF-M.
We could make it more generic by saying that when generating keys the key material is never exposed to the application directly. Even without TF-M only the PSA Crypto implementation would handle the key material. And in the best case the key material could only reside in HW (e.g. KMU keys pushed directly to CRACEN).

For more information, see `Key management functions`_ in the PSA Crypto API documentation.

Configuration
-------------

> For information about how to configure the PSA Crypto API in the |NCS|, see :ref:`psa_crypto_support`.

Reviewer: IMO the title of `psa_crypto_support` should be renamed to Configuring **the** PSA Crypto API**~s~**. (And looks like some text in there could use some corrections, but this is out of scope.)

Usage example
-------------

For usage examples, see :ref:`crypto_samples`, and specifically the :ref:`Persistent key usage <crypto_persistent_key>` sample.

.. _key_storage_huk:

Hardware Unique Key (HUK)
==========================

:ref:`Hardware Unique Keys (HUKs) <lib_hw_unique_key>` are device-specific keys that you can use for key derivation.
HUKs provide a foundation for creating device-unique cryptographic keys without storing key material in software.

This option offers the following features:

* Device uniqueness - Each device has a unique set of HUK values.
* Hardware protected - Keys are stored in hardware-protected memory.

> * Key derivation - HUK is used to derive other keys rather than directly perform cryptographic operations.

Reviewer: ```suggestion
* Key derivation - HUKs are used to derive other keys rather than directly perform cryptographic operations.
```

>   Keys are then passed to the application, which does not use the HUK.

Reviewer: ```suggestion
  The derived keys are then passed to the application.
```

     - :ref:`key_storage_kmu`
     - Recommended for nRF91 Series devices and nRF5340
   * - Master Key for External Storage (MEXT)
     - Deriving keys for encrypting data in external storage (flash) or :ref:`Protected Storage <ug_psa_certified_api_overview_secstorage>`
     - :ref:`key_storage_kmu`
     - Recommended for nRF91 Series devices and nRF5340
   * - CRACEN Isolated Key Generator (IKG)
     - Deriving special hardware keys for CRACEN operations
     - :ref:`key_storage_kmu`
     - | - Regenerated on each CRACEN power cycle from the IKG seed
       | - Supported on nRF54L Series devices; see :ref:`ug_nrf54l_crypto_cracen_ikg`
   * - Device Root Key (KDR)
     - Deriving general-purpose keys
     - Non-volatile memory locked by `ACL <nRF5240 Access Control Lists_>`_
     - | - Should be written to CryptoCell by the bootloader
       | - Supported on and recommended for nRF52840

Device support
--------------

Different nRF devices support different types of HUKs.
See :ref:`lib_hw_unique_key` for device-specific support information.

Usage
-----

> Do not use HUKs directly.

Reviewer: I mean I don't know if you even could?

> Instead, consider using the :ref:`PSA Protected Storage APIs <ug_psa_certified_api_overview_secstorage>` to use HUKs.

Reviewer: I'd rather say either derive keys yourself and use them as volatile keys or store keys through the PSA Crypto API (and not PS/ITS) which under the hood will use the HUK library to encrypt the key material into ITS.

Configuration
-------------

To use the HUK storage options for keys, enable the :kconfig:option:`CONFIG_HW_UNIQUE_KEY` Kconfig option.
For more configuration options, see the :ref:`lib_hw_unique_key` library.

Usage example
-------------

> If you use the HUK library, ensure that the HUK is generated.

Reviewer: ```suggestion
If you use the HUK library, ensure that the HUKs are written.
```

See the :ref:`TF-M Provisioning image <provisioning_image>` sample or the :ref:`Crypto persistent key usage <crypto_persistent_key>` sample for an example.

For an example on how to derive keys from HUKs, see the :ref:`Hardware unique key sample <hw_unique_key_usage>`.

.. _key_storage_iak:

Initial Attestation Key (IAK) / Identity Key
=============================================

The :ref:`Initial Attestation Key (IAK) <lib_identity_key>`, also known as Identity Key, is required by PSA Certified Security Framework for :ref:`ug_psa_certified_api_overview_attestation`.
This key is used to sign attestation tokens that prove the device's identity and integrity state.

This option offers the following features:

* Device identity - Provides cryptographic proof of device identity.
* Attestation support - Required for PSA Attestation implementation.

> * Hardware protection - Encrypted using MKEK and stored in :ref:`data_storage_psa_protected` (when using TF-M) or in Zephyr's :ref:`secure_storage` subsystem or the :ref:`trusted_storage_readme` library (when not using TF-M).

Reviewer: In any way this would be the PSA Protected Storage so let's reduce this bulky sentence to:
```suggestion
* Hardware protection - Encrypted using MKEK and stored in :ref:`data_storage_psa_protected`.
```
But I actually don't know if this is true. @Vge0rge @frkv ?

Reviewer: At least for the case of the nRF54L15 (and similar devices) the IAK is only derived every time, not stored.

.. _key_storage_kmu:

Key Management Unit (KMU)
=========================

The Key Management Unit (KMU) is a hardware peripheral for secure key storage available on select nRF devices.
It provides hardware-level protection for cryptographic keys.

This option offers the following features:

* Hardware security - Keys are stored in dedicated hardware.
* CPU isolation - CryptoCell and CRACEN can use KMU keys without CPU access to key material.
* Metadata only - CPU only knows key metadata, not key material.
* Multiple key slots - Supports storing multiple keys separately.

Device support
--------------

KMU is available on the following devices:

* `nRF5340 <nRF5340 Key management unit_>`_
* nRF91 Series devices, such as `nRF9160 <nRF9160 Key management unit_>`_ or `nRF9151 <nRF9151 Key management unit_>`_
* nRF54L Series devices, such as `nRF54L15 <nRF54L15 Key management unit_>`_

  .. note::
     nRF54L devices are equipped with a KMU that works with the CRACEN peripheral.
     For more information, see :ref:`ug_nrf54l_developing_basics_kmu`.

Key storage
-----------

The KMU stores various types of keys:

* Hardware Unique Keys (MKEK, MEXT)

> * Initial Attestation Key

Reviewer: This seems contradictory with some information above for the nRF54L devices ? (IAK is only derived)

HUKs provide a foundation for creating device-unique cryptographic keys without storing key material in software.

This option offers the following features:

* Device uniqueness - Each device has a unique set of HUK values.
* Hardware protected - Keys are stored in hardware-protected memory.
* Key derivation - Used to derive other keys rather than direct cryptographic operations.
* No key material exposure - Application works with derived keys, not the HUK itself.

Types of HUKs
-------------

The |NCS| supports the following types of HUKs:

.. list-table:: Hardware Unique Key (HUK) Types
   :widths: auto
   :header-rows: 1

   * - HUK type
     - Purpose
     - Storage location
     - Additional notes
   * - Master Key Encryption Key (MKEK)
     - Deriving Key Encryption Keys (KEKs) for encrypting :ref:`Internal Trusted Storage (ITS) <ug_psa_certified_api_overview_secstorage>`
     - :ref:`key_storage_kmu`
     - Recommended for nRF91 Series devices and nRF5340
   * - Master Key for External Storage (MEXT)
     - Deriving keys for encrypting data in external storage (flash) or :ref:`Protected Storage <ug_psa_certified_api_overview_secstorage>`
     - :ref:`key_storage_kmu`

>      - Recommended for nRF91 Series devices and nRF5340

Reviewer: Yeah I think most of this table should rather be in the HUK library page. Here we could just advertize that the HUK library allows generating device-unique keys on the fly using labels and this way they don't even need to be stored.

--------------

Different nRF devices support different types of HUKs.
See :ref:`lib_hw_unique_key` for device-specific support information.

Usage
-----

Do not use HUKs directly.
Instead, derive keys from HUKs using device-specific labels:

1. Provide a label to identify the derived key purpose.
2. Use key derivation functions with the HUK and label to obtain the derived key.

The same combination of label and HUK always produces the same derived key.

Keys can be derived from each HUK, and each HUK can have multiple different labels.
Therefore, each HUK can have multiple different derived keys.

Configuration
-------------

To use the HUK storage options for keys, set the :kconfig:option:`CONFIG_HW_UNIQUE_KEY` Kconfig option.
For more configuration options, see the :ref:`lib_hw_unique_key` library.

Usage example
-------------

If you are using TF-M and it needs the HUK, you must ensure that the HUK is generated.
See the :ref:`TF-M Provisioning image <provisioning_image>` sample for an example.

For an example on how to derive keys from HUKs, see the :ref:`Hardware unique key sample <hw_unique_key_usage>`.

.. _key_storage_iak:

Initial Attestation Key (IAK) / Identity Key

> =============================================

Reviewer: Does this really belong to key storage documentation?

Author: You need to be a bit more specific.

Reviewer: I just don't quite get how the IAK or identity key library is relevant to key storage. On this page we're rather talking about how to store your own key material?
I mean, I kinda see the point, why we might want to talk about this option in this documentation, but saying it's an option to store keys seems a bit wrong.

Reviewer 2: The blog was intended as a way to tie together the different storage and key storage methods in NCS. Because users saw like 5 different ones and though "which is right for me?". Therefore, I mentioned IAK as well, seeing as we have(had?) [Identity key generation](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/2.4.0/nrf/samples/keys/identity_key_generation/README.html) and [Identity key usage](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/2.4.0/nrf/samples/keys/identity_key_usage/README.html) samples, which are easy to find if you search for "NCS key help google"

Author: Moved this section further down under "Additional security considerations", which deals with related topics, but not storage itself.

Hardware Flash Write Protection
===============================

The |NCS| provides the :ref:`fprotect driver <fprotect_readme>` for hardware-level flash protection.
This driver uses underlying hardware peripherals (BPROT, ACL, or SPU) to protect flash regions from unauthorized writes or reads.

Use fprotect to protect stored keys from modification, secure bootloader regions, and implement additional layers of security.

Configuration
-------------

To use the fprotect driver, enable the :kconfig:option:`CONFIG_FPROTECT` Kconfig option.
For more information, see the :ref:`fprotect_readme` page.

.. _key_storage_approtect:

Access Port Protection (AP-Protect)
===================================

The :ref:`access port protection mechanism <app_approtect>` (AP-Protect) prevents the debugger from accessing device resources, including keys and sensitive data.
When enabled, AP-Protect blocks read and write access to CPU registers and memory-mapped addresses.

Nordic Semiconductor recommends enabling AP-Protect in production devices to prevent extraction of keys and sensitive data through debug interfaces.

Configuration
-------------

AP-Protect configuration is device-specific.

> See :ref:`app_approtect_implementation_overview` for detailed information.

Author: I'm linking to the title in the first sentence of the section. Here I link to the implementation details, which happen to be the first section (but this might change in the future).
